File: frontend/src/types/tests14.ts
/** Domain types for tests / LabTestRecord14 */
export type LabTestRecord14Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface LabTestRecord14 {
  id: string;
  externalCode: string;
  displayName: string;
  status: LabTestRecord14Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface LabTestRecord14CreateRequest {
  externalCode: string;
  displayName: string;
  status: LabTestRecord14Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface LabTestRecord14ListResponse {
  items: LabTestRecord14[];
  total?: number;
}

export interface LabTestRecord14StatsResponse {
  activeCount: number;
}

export function isLabTestRecord14Active(rec: LabTestRecord14): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatLabTestRecord14Label(rec: LabTestRecord14): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareLabTestRecord14ByPriority(a: LabTestRecord14, b: LabTestRecord14): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}

File: backend/internal/service/tests_14.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// LabTestRecord14Service encapsulates business rules for LabTestRecord14.
type LabTestRecord14Service struct {
	repo *repository.LabTestRecord14Repository
}

// NewLabTestRecord14Service wires the service.
func NewLabTestRecord14Service(repo *repository.LabTestRecord14Repository) *LabTestRecord14Service {
	return &LabTestRecord14Service{repo: repo}
}

// Create validates and persists a new LabTestRecord14.
func (s *LabTestRecord14Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.LabTestRecord14, error) {
	now := time.Now().UTC()
	rec := &models.LabTestRecord14{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a LabTestRecord14 by id.
func (s *LabTestRecord14Service) Get(ctx context.Context, id string) (*models.LabTestRecord14, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped LabTestRecord14 pages.
func (s *LabTestRecord14Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.LabTestRecord14, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing LabTestRecord14.
func (s *LabTestRecord14Service) Update(ctx context.Context, actor string, rec *models.LabTestRecord14) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.UpdatedBy = actor
	rec.UpdatedAt = time.Now().UTC()
	// repository Update applies optimistic locking and increments version
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a LabTestRecord14.
func (s *LabTestRecord14Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds LabTestRecord14 by external code prefix.
func (s *LabTestRecord14Service) Search(ctx context.Context, facilityID, prefix string) ([]models.LabTestRecord14, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *LabTestRecord14Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}
